### ISSUE TITLE
Tilemap Error using AStar2D

### ISSUE THREAD

**Issue by git2013vb** [godotengine/godot#52321]
### Godot version

v3.4.beta4.mono.official [6a058cbf3]

### System information

Debian 11

### Issue description

When I add an collision shape in tilemap tile the editor allow me to add other collisions on top the first one.
I noticed it because I had this error in console 

```
E 0:00:00.854   add_point: Condition "p_id < 0" is true.
  <C++ Source>  core/math/a_star.cpp:50 @ add_point()
  <Stack Trace> :0 @ void Godot.NativeCalls.godot_icall_3_43(IntPtr , IntPtr , Int32 , Godot.Vector2& , Single )()
                AStar2D.cs:71 @ void Godot.AStar2D.AddPoint(Int32 , Godot.Vector2 , Single )()
                Ground.cs:129 @ Godot.Collections.Array`1[Godot.Vector2] DesertEdge.Ground.AstarAddWalkableCells(Godot.Collections.Array )()
                Ground.cs:89 @ void DesertEdge.Ground._Ready()()

```
And I went in edito tilemap and I tried to delete the collision shape. At this point I noticed that there was another layer underneath - width different colour - and another one and so on.
I was in this status all the time:

![image](https://user-images.githubusercontent.com/13235119/131713206-36044419-e34a-45b7-b810-bfd382f31e46.png)

Then I tried to clean the mess and I deleted all collision shapes, occlusions, regions, and even all tiles...
The error in console still appear and in the scene file I still have all tilemap data: 

```
[node name="Ground" type="TileMap" parent="."]
tile_set = SubResource( 1 )
cell_size = Vector2( 128, 128 )
show_collision = true
collision_layer = 0
collision_mask = 16
format = 1
tile_data = PoolIntArray( 0, 1073741833, 0, 1, 1073741835, 0, 2, 1073741835, 0, 3, 1073741835, 0, 4, 1073741835, 0, 5, 1073741835, 0, 6, 1073741835, 0, 7, 1073741835, 0, 8, 1073741835, 0, 9, 1073741835, 0, 10, 1073741835, 0, 11, 1073741835, 0, 12, 1073741835, 0, 13, 1073741835, 0, 14, 1073741835, 0, 15, 1073741835, 0, 16, 1073741835, 0, 17, 1073741835, 0, 18, 1073741835, 0, 19, 1073741835, 0, 20, 1073741835, 0, 21, 1073741835, 0, 22, 1073741835, 0, 23, 1073741835, 0, 24, 1073741835, 0, 25, 1073741835, 0, 26, 1073741835, 0, 27, 1073741835, 0, 28, 1073741835, 0, 29, 1073741835, 0, 30, 1073741835, 0, 31, 1073741835, 0, 32, 1073741835, 0, 33, 1073741835, 0, 34, 1073741835, 0, 35, 1073741835, 0, 36, 1073741835, 0, 37, 1073741835, 0, 38, 1073741835, 0, 39, 1073741835, 0, 40, 1073741835, 0, 41, 1073741835, 0, 42, 1073741835, 0, 43, 1073741835, 0, 44, 1073741835, 0, 45, 1073741835, 0, 46, 1073741835, 0, 47, 1073741835, 0, 48, 1073741835, 0, 49, 1073741835, 0, 50, 1073741835, 0, 51, 1073741835, 0, 52, 1073741835, 0, 53, 1073741835, 0, 54, 1073741835, 0, 55, 1073741835, 0, 56, 1073741835, 0, 57, 1073741835, 0, 58, 1073741835, 0, 59, 1073741835, 0, 60, 1073741835, 0, 61, 1073741835, 0, 62, 1073741835, 0, 63, 1073741835, 0, 64, 1073741835, 0, 65, 1073741835, 0, 66, 1073741835, 0, 67, 1073741835, 0, 68, 1073741835, 0, 69, 
```

while I don't have anything in design:
![image](https://user-images.githubusercontent.com/13235119/131713741-1d0718c0-b843-402f-a4a6-2690e55acfa7.png)
![image](https://user-images.githubusercontent.com/13235119/131713837-235c4344-179f-4185-839c-60b2a459ed1b.png)

### Steps to reproduce

See before

### Minimal reproduction project

NA

**Comment by Calinou** [comment]
@git2013vb Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by git2013vb** [comment]
[TestTileSet.zip](https://github.com/godotengine/godot/files/7096243/TestTileSet.zip)
Also if you try to point-mouse-click in this section:

![image](https://user-images.githubusercontent.com/13235119/131785709-98c370e9-fb8b-4e1a-83bd-4c56ce58dfd7.png)

The asta2d behave weird like:
![image](https://user-images.githubusercontent.com/13235119/131785773-131b5093-0fd1-4827-a5c2-282f131b1976.png)

**Comment by git2013vb** [comment]
Here:
[TestTileSet.zip](https://github.com/godotengine/godot/files/7096337/TestTileSet.zip)

The AStar2D behave weird in some cases like if you try to point to the road on the right side it will set a weird path:
 
![image](https://user-images.githubusercontent.com/13235119/131787706-7470df7c-301a-4556-a88c-4bea2806951f.png)
![image](https://user-images.githubusercontent.com/13235119/131787794-1ae0370b-0fac-40ae-bd4e-371e449f6a11.png)
![image](https://user-images.githubusercontent.com/13235119/131787881-55f1f7cc-dbbb-47ce-81a4-123e4ec121cc.png)

**Comment by kleonc** [comment]
> When I add an collision shape in tilemap tile the editor allow me to add other collisions on top the first one.
> I noticed it because I had this error in console
> 
> ```
> E 0:00:00.854   add_point: Condition "p_id < 0" is true.
>   <C++ Source>  core/math/a_star.cpp:50 @ add_point()
>   <Stack Trace> :0 @ void Godot.NativeCalls.godot_icall_3_43(IntPtr , IntPtr , Int32 , Godot.Vector2& , Single )()
>                 AStar2D.cs:71 @ void Godot.AStar2D.AddPoint(Int32 , Godot.Vector2 , Single )()
>                 Ground.cs:129 @ Godot.Collections.Array`1[Godot.Vector2] DesertEdge.Ground.AstarAddWalkableCells(Godot.Collections.Array )()
>                 Ground.cs:89 @ void DesertEdge.Ground._Ready()()
> ```

This error is trigerred by you trying to add a point with negative index to `AStar2D`, it has nothing to do with the tiles and their collision shapes. I guess some understandable error message needs to be added in here:
https://github.com/godotengine/godot/blob/d4a84be4fb91ef425b4ca55f3d24464bbd9906c9/core/math/a_star.cpp#L49-L52

---

> And I went in edito tilemap and I tried to delete the collision shape. At this point I noticed that there was another layer underneath - width different colour - and another one and so on.

Can't notice anything like this. I can see you have a collision shape set for 8 of your tiles, the same data is in the scene file, everything looks just fine:
![SldKaAcUqX](https://user-images.githubusercontent.com/9283098/131835052-07475b21-dc14-48c6-b2a6-874cd7fc0235.gif)
```
(...)
[sub_resource type="ConvexPolygonShape2D" id=2]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=3]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=4]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=5]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=6]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=7]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=8]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="ConvexPolygonShape2D" id=9]
points = PoolVector2Array( 128, 128, 0, 128, 0, 0, 128, 0 )

[sub_resource type="TileSet" id=10]
(...)
36/shapes = [ {
"autotile_coord": Vector2( 0, 0 ),
"one_way": false,
"one_way_margin": 1.0,
"shape": SubResource( 2 ),
"shape_transform": Transform2D( 1, 0, 0, 1, 0, 0 )
} ]
(...) // and so on for the other tiles
```
Similarly can't see a problem with deleting collision shapes:
![vsXnGKZEr9](https://user-images.githubusercontent.com/9283098/131835773-a2c8e442-6dcc-4e3f-871d-460693c3c393.gif)


To sum up: I don't see a single bug in here, just some buggy/incorrect code of yours, @git2013vb .

**Comment by git2013vb** [comment]
>Can't notice anything like this.
>Similarly can't see a problem with deleting collision shapes:
You can't them any-more. I deleted the old map. (was a map created with godot version 3 that is supposed to be compatible right?)

>To sum up: I don't see a single bug in here,
Its not to early to say that? How to explain the wrong path that happen also in empty map?
I documentation there is no warning/instructions/examples about not to make a map in negative coordinates. Coordinate that are allowed to use) 
The usually way people use the editor is not what developer are used to test and to think  it will be used. Without a clear guideline people like myself are going to delete/add/rename/move (repeat again in not same order) any map parts.  Also how about non documented API?  How I 'm supposed to use them in correct way if  there is just a syntax ? I can read other people mind? Sometime I can but not all the time. :)
last thing: How about the weird path mentioned from my screenshots? It happen even without any tile on map.
Thanks.
ps: any eta to have a more clear interface?

**Comment by git2013vb** [comment]
Now i setup the map starting from 0,0  but I have another error:
```
E 0:00:05.321   get_point_path: Condition "!to_exists" is true. Returned: PoolVector<Vector2>()
  <C++ Source>  core/math/a_star.cpp:700 @ get_point_path()
  <Stack Trace> :0 @ Godot.Vector2[] Godot.NativeCalls.godot_icall_2_50(IntPtr , IntPtr , Int32 , Int32 )()
                AStar2D.cs:339 @ Godot.Vector2[] Godot.AStar2D.GetPointPath(Int32 , Int32 )()
                Ground.cs:231 @ void DesertEdge.Ground.RecalculatePath()()
                Ground.cs:38 @ void DesertEdge.Ground.set_PathStartPosition(Godot.Vector2 )()
                Ground.cs:204 @ Godot.Collections.Array`1[Godot.Vector2] DesertEdge.Ground.GetAstarPath(Godot.Vector2 , Godot.Vector2 )()
                Ground.cs:108 @ Godot.Collections.Array`1[Godot.Vector2] DesertEdge.Ground.SetNewDestination(Godot.Vector2 , Godot.Vector2 )()
                :0 @ System.Object Godot.NativeCalls.godot_icall_2_537(IntPtr , IntPtr , System.String , System.Object[] )()
                Object.cs:377 @ System.Object Godot.Object.Call(System.String , System.Object[] )()
                Peer.cs:76 @ void DesertEdge.Peer.ChangeState(DesertEdge.Peer+STATES )()
                Peer.cs:66 @ void DesertEdge.Peer._UnhandledInput(Godot.InputEvent )()
```
the point found are:
```

(749195) 17:05:26.877  :DEBUG	Borderland[Void SpawnMyPlayer(Godot.Vector2, Godot.Vector2, Int32)] 
(749195) 17:05:26.884  :DEBUG	Borderland[Void SpawnMyPlayer(Godot.Vector2, Godot.Vector2, Int32)] Spawned peer 0 at (200, 200) dest (200, 200)
(749195) 17:05:31.266  :DEBUG	Ground[Void RecalculatePath()] 78 0
(749195) 17:05:31.270  :DEBUG	Ground[Void RecalculatePath()] 78 1493
(749195) 17:05:31.270  :DEBUG	Ground[Void RecalculatePath()] 78 1493
 Cannot open file '/root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AStar2D.cs'.
 Failed to read file: '/root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AStar2D.cs'.
```
this means I need to start the TileMap from 1,1 and not from 0,0?

**Comment by kleonc** [comment]
> You can't see them any-more. I deleted the old map. (was a map created with godot version 3 that is supposed to be compatible right?)

Which exact version? Personally I'm not sure whether there are some compatibility breakages between some `3.x` versions but maybe someone else could clarify it for you if you'd provide more details.
So yeah, maybe you've indeed faced some bug, maybe not, I simply don't know. I haven't seen it in action and I don't know how to reproduce it. Without a way to reproduce it it's really hard to track it down. :wink: So if you have any more details about what exactly happend and what steps leaded to it then please provide such details (for example in your original post you didn't say a word about migrating from some older version).



>>   To sum up: I don't see a single bug in here,

>Its not to early to say that?

No, currently I don't see a single bug in here. Nothing more, nothing less.

> How to explain the wrong path that happen also in empty map?
Or a normal map (just positive tiles) that have the errors?
(...)
> last thing: How about the weird path mentioned from my screenshots? It happen even without any tile on map.

Incorrect logic in your code. Here you go, your project "a little" modified: [TestTileSet.zip](https://github.com/godotengine/godot/files/7100128/TestTileSet.zip), I see no errors. I think the main problem with your logic was incorrect in-`AStar2D` point id calculation from the in-`TileMap` cell coords. But that's not a place to fix the issues with your code logic. This is a bug tracker. If you have problems with your code (or if you're not really sure if you're facing an in-engine bug) then visit some of the [community channels](https://godotengine.org/community), there are plenty of helpful people in there.

> I documentation there is no warning/instructions/examples about not to make a map in negative coordinates. Coordinate that are allowed to use)

 - [AStar2D.add_point](https://docs.godotengine.org/en/3.4/classes/class_astar2d.html#class-astar2d-method-add-point): "_(...) The id must be 0 or larger_, (...)"
 ![Code_kQLJGct5Go](https://user-images.githubusercontent.com/9283098/131877205-707730b3-9874-44b1-9c06-22d5d01fe407.png)

> The usually way people use the editor is not what developer are used to test and to think it will be used. Without a clear guideline people like myself are going to delete/add/rename/move (repeat again in not same order) any map parts.

Seems like for some reason you think that `AStar2D` and `TileMap` depends on each other. They're of course independent of each other, I have no idea why would anyone think otherwise. If there's some dependency between these two, it's because of your own code.

>  Also how about non documented API? How I 'm supposed to use them in correct way if there is just a syntax ? I can read other people mind? Sometime I can but not all the time. :)

What non documented API? Please be more specific.
If you're talking about online C# API documentation then I have no idea if it's planned or (if it's planned) what's the current state of it. Currently you've got [this section in the docs](https://docs.godotengine.org/en/3.4/getting_started/scripting/c_sharp/index.html) and in IDE autocompletion/docs. Also you can always refer to [GDScript's API](https://docs.godotengine.org/en/3.4/classes/index.html), it's mostly the same.

> You show me some source code in c++. I barely work in C#... How I'm supposed to understand that one without the context/knowledge ? To use this game engine I need to be an expert like yourself? How you assume I'm that?

Well that part wasn't addressed directly to you, rather to anyone reading it (even to myself :smile: ). I guess you could've interpret it this way, my bad.

> ps: any eta to have a more clear interface?

No idea what you're asking about.

> Also where are the ID that I need when I use this method? GetUsedCellsById

Well, that's tile ID. I guess the answer you're expecting (not sure about this) is: in the `TileSet`. Anyway, this question doesn't really belong in here. How is it related to the issue you've reported? As I've mentioned there are community channels for such questions. And if you want to report an issue with the documentation then please do it, but it's not the place for it:
![fREbnzvlX9](https://user-images.githubusercontent.com/9283098/131886352-6f0d3270-2624-4fcb-9f2f-364dee0a191a.png)


> Seems like there are tons of info that are not allowed to be know to common users - its a military secret? I feel mistreated. And is not the first time.

Again, that's just offtopic. I kinda feel mistreated because you're not really talking about your bug in here.

**Comment by git2013vb** [comment]
>Again, that's just offtopic. I kinda feel mistreated because you're not really talking about your bug in here.

Apologise.
You right. I was off topic. And yes I was "porting" a map from an old version ( still same main release - 3v - where was possible to set negative numbers). So back the time my logic was sound. Also  agree about documentation that is not the right place.

Back on track. Last zip file contain a project that show an error related to AStar2D. Did you take a look? It don't seem to me you mentioned anything related to it. It is a bug?
Thanks

**Comment by kleonc** [comment]
> ( still same main release - 3v - where was possible to set negative numbers)

I've just checked it and the check disallowing negative ids in `AStar::add_point` method is in the source code [since `AStar` was added](https://github.com/godotengine/godot/commit/827a9aa8294e7e2405f645579cc3e7044f3be079#diff-c5cef4bba4136dd193065674214d428352b856d60a9d10f2195d2fcd1e162167R15)... which hapened almost 5 years ago (something between `2.0` and `2.1` versions if I've checked correctly). So as of today points with negative ids has been disallowed in `Astar` (and `AStar2D`) in all `3.x` versions.
I suspect you're talking about cell coords in the `TileMap`, not about ids in the `AStar2D`. As I've already mentioned it in my previous comment, these are unrelated. And cell coords in the `TileMap` can be negative, yes.

> Back on track. Last zip file contain a project that show an error related to AStar2D. Did you take a look? It don't seem to me you mentioned anything related to it. It is a bug?

I indeed didn't take a look at it earlier. I've assumed the main issue here is your code/logic because I've already seen there was no errors after I've modified your project a little (have you checked out that modified project I've linked in my previous comment?). Now I've looked at that last zip of yours and I was right, the issue is, again, your logic:
- Your in editor map size (in cells) is (78, 78), not (77, 77) which is hard coded by you. Your first cell is (0, 0), not (1, 1).
- In `AstarAddWalkableCells` method for both `x` and `y` you're iterating from 1 instead of from 0. Again, your first cell is (0, 0), not (1, 1).

So yeah, this is an issue you could have easily solved by asking for help in some community channel, not here (and you'd most likely solve it faster that way).

As of now, I'd say some error messages in `AStar` should be made easier to understand and that's all. I'll later open a PR for it (and I'll ping you so you could give some feedback about added error messages).

**Comment by git2013vb** [comment]
Feel free to close this issue when you think is the right time. :)
